File: Python/graph_algorithms/09_negative_cycles.py
import sys
from collections import defaultdict


class DirectedGraph(object):
    def __init__(self, edge_list, nr_nodes):
        self._adjacency_list = self._convert_edgelist_to_adjacency_list(edge_list)
        self._nr_nodes = nr_nodes

    @property
    def adjacency_list(self):
        return self._adjacency_list

    @property
    def nr_nodes(self):
        return self._nr_nodes

    def _convert_edgelist_to_adjacency_list(self, edge_list):
        adjacency_list = defaultdict(list)
        for node_from, node_to, weight in edge_list:
            adjacency_list[node_from].append((node_to, weight))
        return dict(adjacency_list)


def update_distances(graph, distances):
    update_happened = False
    for node, neighbours in graph.adjacency_list.items():
        dist_node = distances[node-1]
        for neighbour, weight in neighbours:
            distance_through_node = dist_node + weight
            if distances[neighbour-1] > distance_through_node:
                distances[neighbour-1] = distance_through_node
                update_happened = True
    return update_happened


def has_negative_cycle(graph):
    n = graph.nr_nodes
    distances = [0] * n
    complete_pass_on_edges = 1
    while complete_pass_on_edges <= n:
        is_some_edge_updated = update_distances(graph, distances)
        if complete_pass_on_edges == n and is_some_edge_updated:
            return True
        complete_pass_on_edges += 1
    return False


def detect_negative_cycle(edge_list, nr_nodes):
    graph = DirectedGraph(edge_list, nr_nodes)
    return has_negative_cycle(graph)


def convert_to_intlist(line):
    return [int(elem) for elem in line.split()]


def read_weighted_edges(reader, nr_edges):
    edges = []
    for _ in range(nr_edges):
        edge = tuple(convert_to_intlist(next(reader)))
        edges.append(edge)
    return edges


def main():
    reader = sys.stdin
    nr_nodes, nr_edges = convert_to_intlist(next(reader))
    edge_list = read_weighted_edges(reader, nr_edges)
    result = detect_negative_cycle(edge_list, nr_nodes)
    print(int(result))


if __name__ == "__main__":
    main()

File: Rust/graph_algorithms/src/bin/09_negative_cycles.rs
use std::collections::HashMap;
use std::io;

#[derive(Clone)]
struct Edge {
    from: i32,
    to: i32,
    weight: i32,
}

struct DirectedGraph {
    nr_nodes: i32,
    adjacency_list: HashMap<i32, Vec<Edge>>
}

impl DirectedGraph {
    fn from_edge_list(nr_nodes: i32, edges: Vec<Edge>) -> Self {
        let mut adjacency_list = HashMap::new();
        for edge @ Edge { from, ..} in edges {
            adjacency_list.entry(from).or_insert_with(Vec::new).push(edge);
        }

        DirectedGraph { nr_nodes, adjacency_list }
    }
}

fn convert_to_int_vector(line: &str) -> Vec<i32> {
    line
        .split_whitespace()
        .map(|x| x.parse().unwrap())
        .collect()
}

fn read_line() -> String {
    let mut line = String::new();
    io::stdin().read_line(&mut line).unwrap();

    line
}

fn read_pair() -> (i32, i32) {
    let line = read_line();
    let xs: Vec<i32> = convert_to_int_vector(&line);
    let [a, b] = xs.as_slice() else {
        panic!("Malformed input: {}", line);
    };
    (*a, *b)
}

fn read_edges(nr_edges: usize) -> Vec<Edge> {
    let mut edges = Vec::with_capacity(nr_edges);
    for _ in 0..nr_edges {
        let line = read_line();
        let xs: Vec<i32> = convert_to_int_vector(&line);
        let [a, b, w] = xs.as_slice() else {
            panic!("Malformed input: {}", line);
        };
        edges.push(Edge { from: *a, to: *b, weight: *w });
    }

    edges
}

fn update_distances(graph: &DirectedGraph, distances: &mut [i32]) -> bool {
    let mut any_node_updated = false;
    for (node, neighbors) in &graph.adjacency_list {
        let dist_node = distances[*node as usize - 1];
        for &Edge {to, weight, ..} in neighbors {
            let distance_through_node = dist_node + weight;
            if distances[to as usize - 1] > distance_through_node {
                distances[to as usize - 1] = distance_through_node;
                any_node_updated = true
            }
        }
    };
    any_node_updated
}

fn has_negative_cycle(graph: DirectedGraph) -> bool {
    let n = graph.nr_nodes as usize;
    let mut distances = vec![0; n];
    let mut complete_pass_on_edges = 1;
    while complete_pass_on_edges <= n {
        let has_anything_updated = update_distances(&graph, &mut distances);
        if complete_pass_on_edges == n && has_anything_updated {
            return true
        }
        complete_pass_on_edges += 1;
    }
    false
}

fn main() {
    let (nr_nodes, nr_edges) = read_pair();
    let edge_list = read_edges(nr_edges as usize);
    let graph = DirectedGraph::from_edge_list(nr_nodes, edge_list);
    let result = has_negative_cycle(graph);
    println!("{}", result as i32);
}
